docs/training/apis-howto/deepspeed/autotuning.rst

```diff
@@ -0,0 +1,247 @@
+settings through an easy-to-use API with very few changes required in user-code, as we describe
+in the remainder of this user guide. ``dsat`` can be used with any of
+:class:`~determined.pytorch.deepspeed.DeepSpeedTrial`, Core API, and HuggingFace Trainer.
+ 
+
+
+What to Expect
+================
+
+Assuming you have DeepSpeed code which already functions, autotuning is as easy as inserting one or
+two helper functions into your code and then replacing the usual experiment launching command 
+
+.. code::
+
+  det experiment create deepspeed.yaml .
+
+with
+
+.. code::
+
+  python3 -m determined.pytorch.dsat asha deepspeed.yaml .
+
+where ``deepspeed.yaml`` is the Determined configuration for a ``single`` experiment. There is no
+need to write a special configuration file to use ``dsat``.
+
+The above uses the ASHA algorithm (TODO: link to ASHA page) to tune the DS parameters and is one
+of the three search methods we currently provide:
+
+* ``asha``: adaptively searches over randomly selected DeepSpeed configurations, yielding more
+   compute resources to well-performing classes of configurations.
+
+* ``binary``: simple binary searches over the batch size for randomly-generated DS configurations.
+
+* ``random``: a search over random DeepSpeed configurations with an aggressive early-stopping
+  criteria based on domain-knowledge of DeepSpeed and the search history.
+
+DeepSpeed Autotune is built on top of Custom Searcher (TODO: link) 
+which starts up two separate experiments:
+
+* A ``single`` search runner experiment whose role is to coordinate and schedule the actual trials
+  which run model code. 
+* A ``custom`` experiment which contains the trials referenced above whose results are reported back
+  to the search runner.
+
+The logs of the ``single`` search runner experiment will contain information regarding the size of the model,
+the GPU memory available, the activation memory required per example, and an approximate computation
+of the maximum batch size per zero stage (which is used to guide the starting point for all
+autotuning searches). When a best-performing DS configuration is found, the
+corresponding ``json`` configuration file will be written to the search runner's checkpoint directory,
+along with a file detailing the configuration's corresponding metrics.
+
+The ``custom`` experiment instead holds the visualization and summary tables which outline the results for
+every Trial. Initially, a one-step profiling Trial is created to gather the above information regarding
+the model and available hardware. Subsequently, multiple short Trials are submitted which each report
+back metrics such as ``FLOPS_per_gpu``, ``throughput`` (samples/second), and latency timing information.
+
+In the following sections, we describe the specific user-code changes which must be made if using
+``dsat`` with :class:`~determined.pytorch.deepspeed.DeepSpeedTrial`, Core API, and HuggingFace
+Trainer, respectively.
+
```

Comment:
What to Expect section
- i think this is a good overview and intro to the DeepSpeedTrial, Core API, and HuggingFace Trainer sections
- I renamed it "How the Autotuning Process Works".
- this incorporates some of Taylor's feedback.

**How the Autotuning Process Works**

 You do not need to create a special configuration file to use ``dsat``.
Assuming you already have functional DeepSpeed code, the autotuning process is straightforward. You just need to insert one or two helper functions into your code and replace the usual experiment launching command.
For example, using the ASHA algorithm, you'll replace the following experiment launching command:

.. code::

  det experiment create deepspeed.yaml .
with the following command

.. code::

  python3 -m determined.pytorch.dsat asha deepspeed.yaml .

where ``deepspeed.yaml`` represents the Determined configuration for a ``single`` experiment.

Determined's DeepSpeed Autotune provides three search methods:

- ``asha``: Adaptively searches over randomly selected DeepSpeed configurations, allocating more compute resources to well-performing configurations.
- ``binary``: Performs simple binary searches over the batch size for randomly-generated DS configurations.
- ``random``: Conducts a search over random DeepSpeed configurations with an aggressive early-stopping criteria based on domain-knowledge of DeepSpeed and the search history.

DeepSpeed Autotune is built on top of Custom Searcher, which initiates two separate experiments:

- ``single`` Search Runner Experiment: This experiment coordinates and schedules the trials that run the model code.
- ``custom`` Experiment: This experiment contains the trials referenced above whose results are reported back to the search runner.

**Logs and Metrics**

The logs of the ``single`` search runner experiment contain information regarding the model size, available GPU memory, activation memory required per example, and an approximate computation of the maximum batch size per zero stage. This information guides the starting point for all autotuning searches.
When a best-performing DeepSpeed (DS) configuration is found, the corresponding ``json`` configuration file is written to the search runner's checkpoint directory along with a file that details the configuration's corresponding metrics.

The ``custom`` experiment holds the visualization and summary tables outlining the results for every Trial. Initially, a one-step profiling Trial is created to gather initial profiling information about the model and available hardware. The orchestrator experiment then takes the initial profiling information and creates a series of Trials to search for the most optimal hyperparameters for the experiment. These Trials are submitted with each reporting metrics back such as FLOPS_per_gpu, throughput (samples/second), and latency timing information.

**User-Code Changes**

To use dsat with :class:`~determined.pytorch.deepspeed.DeepSpeedTrial`, Core API, and HuggingFace Trainer, specific changes must be made to your user code. In the following sections, we will describe these changes in detail for each use case.